Nix language basics RITZA edit (#6)
fricklerhandwerk/nix.dev: source/tutorials/nix-language.md

  A Linux distribution that can be configured fully declaratively and is based on Nix and Nixpkgs.

>   Its underlying modular configuration system is written in the Nix language and uses packages from Nixpkgs.

Reviewer: The comma was for emphasis.

- Familiar with functional programming: 1 hour
- Proficient with functional programming: 30 minutes

> We recommend you run all examples.

Reviewer: ```suggestion
We recommend running all examples.
```

Does that fit better?

You may quickly encounter Nix language expressions that look very complicated.

As with any programming language, the required amount of Nix language code closely matches the complexity of the problem it is supposed to solve and reflects how well the problem – and its solution – is understood.

Building software is a complex undertaking, and Nix both *exposes* and *allows managing* this complexity with the Nix language.

The purpose of the Nix language is to create *build tasks*: precise descriptions of how the contents of existing files are used to derive new files.

:::{important}
A build task in Nix is called a *derivation*.
:::

> The Nix language has a few basic constructs which can be combined arbitrarily:

Reviewer: ```suggestion
The Nix language has a small number of basic constructs which can be combined arbitrarily:
```

Does that convey the meaning more precisely? I want to emphasize  "there are not many of them", not colloquially "there are only a couple of them".

(let)=
### `let ... in ...`

> Also known as "`let` expression" or "`let` binding"

Reviewer: The original is typographically correct. It's just more annoying to edit.

    /nix/var/nix/profiles/per-user/root/channels/nixpkgs/lib

> While you will see many such examples, we recommend you [avoid search paths](search-path) in practice, as they are [impurities](impurities), which are not reproducible.

Reviewer: ```suggestion
While you will see many such examples, we recommend [avoiding search paths](search-path) in practice, as they are [impurities](impurities), which are not reproducible.
```

- On its right is the function body.

Function declarations in the Nix language can appear in different forms.

> Each of them is explained below, and here is an overview:

Reviewer: I disagree with saying "below" or "above", because on the web the publisher has no final control over presentation. Anyone is free to display this in multiple columns, pages, backwards, whatever.

#### Multiple arguments

Also known as "[curried] functions".

Nix functions take exactly one argument.
Multiple arguments can be handled by nesting functions.

> Such a nested function can be used like a function that takes multiple arguments but offers additional flexibility.

Reviewer: I find this makes it harder to parse the sentence.

<!-- TODO: find a place for operators -->

> We recommend to at least skim them to familiarize yourself with what is available.

Reviewer: No, we use BE.

A derivation's output path is fully determined by its inputs, which in this case come from *some* version of Nixpkgs.

This is why we recommend [avoiding search paths](search-path), except in examples intended for illustration only.
:::

<details><summary>Detailed explanation</summary>

> The example imports the Nix expression from the search path `<nixpkgs>` and applies the resulting function to an empty attribute set `{}`.

Reviewer: This is a sequence, not parallel. I'm not super strong on English grammar and placing commas, but removing that one feels it would distort the meaning.

To get things done:

> - [Declarative and reproducible developer environments](declarative-reproducible-envs) – create reproducible shell environments from a Nix file

Reviewer: The title will be filled automatically.

- [Garbage Collection](https://nixos.org/manual/nix/stable/package-management/garbage-collection.html) – remove unused build results from the Nix store

To learn more:

If you worked through the examples, you will have noticed that reading the Nix language reveals the structure of the code, but does not necessarily tell what the code actually means.

Often it is not possible to determine from the code at hand:
- the data type of a named value or function argument,

> - the data type a called function accepts for its argument, or

Reviewer: This is not required grammatically and makes the list harder to parse for no good reason.
